### PR TITLE
Replaced usages of EnsimeModule with EnsimeProject. Deleted EnsimeModule

### PR DESCRIPTION
**Pull request by VC1995**
Deleted `EnsimeModule`.
`modules` in `EnsimeConfig` now correspond to a `List[EnsimeProject]`.

I have assumed the following equivalents :
`compileJars` in EnsimeModule <--> `libraryJars` in EnsimeProject
`referenceSourceJars` in EnsimeModule <--> `librarySources` in EnsimeProject
`docJars` in EnsimeModule <--> `libraryDocs` in EnsimeProject
Others are pretty straightforward.

The hack for handing `TypeCheckModule(moduleName)` and `UnloadModule(moduleName)` in `Analyzer.scala` needs review. Using `EnsimeProject` breaks files that would have been included in one `EnsimeModule` into two EnsimeProjects with `compile` and `test` configurations(probably more).  Hence we now need to filter EnsimeProject with `EnsimeProject.id.project == moduleName`, instead of a simple get.

If there's anything wrong, please let me know and I'll try my best to correct it. :)

### CODE REVIEW

File: api/src/main/scala/org/ensime/api/config.scala [hunk 1]
```diff
@@ -29,60 +28,33 @@ final case class EnsimeConfig(
...
-    (javaSources ++ subprojects.flatMap(_.referenceSourceRoots)).toSet
+    (javaSources ++ projects.flatMap(_.librarySources)).toSet
 
   // some marshalling libs (e.g. spray-json) might not like extra vals
-  val modules = subprojects.map { module => (module.name, module) }.toMap
+  val modules = projects.map { module => (module.id, module) }.toMap
 
   def compileClasspath: Set[File] = modules.values.toSet.flatMap {
-    m: EnsimeModule => m.compileDeps ++ m.testDeps
+    m: EnsimeProject => m.libraryJars
```
Comment (yyadavalli) [inline api/src/main/scala/org/ensime/api/config.scala]: @VC1995 with the work you are doing in creating an Analyzer per module, I don't think we will need these any longer.
Same for `targetClasspath`, `allJars`, `allTargets` and `allDocJars`.

File: core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala [hunk 2]
```diff
@@ -52,6 +52,8 @@ object EnsimeConfigFixture {
     )
     System.err.flush()
     sys.exit(1)
+  } else {
+    println(dotEnsime.getAbsolutePath)
```
Comment (yyadavalli) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: Why is this needed here?

File: core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala [hunk 3]
```diff
@@ -141,18 +143,17 @@ object EnsimeConfigFixture {
       if (copyTargets) renameAndCopy(from)
       else rename(from)
 
-    def cloneModule(m: EnsimeModule): EnsimeModule = m.copy(
+    def cloneModule(m: EnsimeProject): EnsimeProject = m.copy(
```
Comment (yyadavalli) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: You might want to rename it to `cloneProject` for consistency

File: core/src/main/scala/org/ensime/core/RichPresentationCompiler.scala [hunk 4]
```diff
@@ -28,7 +28,7 @@
  * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
  * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR
  * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
- * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DsubpAMAGES (INCLUDING, BUT NOT
```
Comment (yyadavalli) [inline core/src/main/scala/org/ensime/core/RichPresentationCompiler.scala]: Committed a typo

File: core/src/main/scala/org/ensime/indexer/SearchService.scala [hunk 5]
```diff
@@ -137,7 +137,7 @@ class SearchService(
     def findBases(): (Set[FileObject], Map[FileName, Set[FileObject]]) = {
       val (jarFiles, dirs) = config.modules.flatMap {
         case (name, m) =>
```
Comment (yyadavalli) [inline core/src/main/scala/org/ensime/indexer/SearchService.scala]: The match is on a `id`, you can rename it for consistency, even though its not relevant here. You can also ignore it with `_`

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 6]
```diff
@@ -185,13 +185,14 @@ class Analyzer(
...
-      config.modules get (moduleName) foreach {
+      //hack: to typecheck both compile and test configurations
+      (config.modules filter { case (moduleId, module) => moduleId.project == moduleName }).values foreach {
         module =>
           val files: Set[SourceFileInfo] = module.scalaSourceFiles.map(s => SourceFileInfo(EnsimeFile(s), None, None))(breakOut)
           sender ! scalaCompiler.handleReloadFiles(files)
       }
     case UnloadModuleReq(moduleName) =>
-      config.modules get (moduleName) foreach {
+      (config.modules filter { case (moduleId, module) => moduleId.project == moduleName }).values foreach {
```
Comment (yyadavalli) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: You can use `filterKeys` to simplify the code a bit. `filterKeys(id => id.project == moduleName)`.

File: core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala (hunk 2 above)
Comment (VC1995) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: I was doing that for testing something. I'm so sorry, I'll correct it.

Comment (yyadavalli) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: Okay, also you could squash your commits (if you like) for a cleaner history.

File: core/src/main/scala/org/ensime/core/Analyzer.scala (hunk 6 above)
Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: Yep, I'll use it instead.

File: api/src/main/scala/org/ensime/api/config.scala (hunk 1 above)
Comment (VC1995) [inline api/src/main/scala/org/ensime/api/config.scala]: Will it be okay if we remove these as we bring in code for Analyzer per module? I am just a little hesistant to remove all of it now as they are being used in a few classes and i am not sure about refactoring them.

Comment (yyadavalli) [inline api/src/main/scala/org/ensime/api/config.scala]: Sure, just wanted to point that out as you do the cleanup.

File: core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala (hunk 2 above)
Comment (VC1995) [inline core/src/it/scala/org/ensime/fixture/EnsimeConfigFixture.scala]: I'll remove the comment on TypeCheckModule(moduleName). Did I leave out more comments?

**Comment by VC1995** [comment]
BasicWorkflow test now fails as I get two void responses on Type checking a module as projects with `EnsimeProjectId("testing_simple","compile")` and `EnsimeProjectId("testing_simple","test")` are loaded.
This was expected. Should I add an extra expectMsg(VoidResponse) statement? or Refactor code to send only a single response?

**Comment by yyadavalli** [comment]
I don't think adding an extra `expectMsg(VoidResponse)` will help. You should either aggregate the result for all modules, or change it to accept a `EnsimeProjectId`, you can assume the default configuration of `compile`, if multiple are found and one is not provided.

**Comment by VC1995** [comment]
For TypeCheckModule
I aggregated the EnsimeServerErrors into one with a description formed by combining all the descriptions inserting newline between them. If no error is obtained I return a Void Response.
For UnloadModule
I simply return a Void Response after unloading all modules with module.id.project == moduleName.

File: core/src/main/scala/org/ensime/core/Analyzer.scala [hunk 7]
```diff
@@ -185,18 +185,29 @@ class Analyzer(
...
       }
     case UnloadModuleReq(moduleName) =>
-      config.modules get (moduleName) foreach {
+      (config.modules filterKeys (_.project == moduleName)).values foreach {
         module =>
           val files = module.scalaSourceFiles.toList
           files.foreach(scalaCompiler.askRemoveDeleted)
-          sender ! VoidResponse
       }
+      sender ! VoidResponse
```
Comment (yyadavalli) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: I think you need to enclose them in a block

Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: That'll send multiple Void Responses. We wanted to send just one, right?

Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: This is kind of a hack though. Currently if we received a request and we had no module named `moduleName`, we wouldn't send back a Void Response. Otherwise, it's same.
Will that be a problem? If so, we can check if we get some EnsimeProject with `module.id.project` equal to moduleName and send back a Void Response only then.
But does this really matter?

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: I think you are fine

Comment (yyadavalli) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: But for `TypeCheckModule` we maybe better off, accepting an `id`, it would be much simpler to change the call site to send the configuration.

Comment (VC1995) [inline core/src/main/scala/org/ensime/core/Analyzer.scala]: Yep. I think so too. I'll make that change. :)

**Comment by VC1995** [comment]
I changed `TypeCheckModule(moduleId)` and `UnloadModuleReq(moduleId)`. They now accept an EnsimeProjectId as parameter.

**Comment by yyadavalli** [comment]
@VC1995 Looks good. We need to watch out for any failures with clients using `TypecheckModule`

**Comment by VC1995** [comment]
Where do I look for making the required changes?
We need to pass an EnsimeProjectId instead of a string.
Also, having an implicit function to convert a `String` to `EnsimeProjectId` with config as `"compile"` in the companion object of `EnsimeProjectId` could work for now, right?

**Comment by yyadavalli** [comment]
I don't think you need to make any changes on the server side. 

The clients such as emacs need to make sure they are passing correct info such that the requests are unmarshalled correctly. So if we ever get issues regarding failure of typecheck module we need to update the client as needed.

We can add an implicit conversion if we are worried about backward compatibility, but we are not. If you are using Developer edition, you can expect to see stuff breaking.

File: .gitignore [hunk 8]
```diff
@@ -8,3 +8,4 @@ ensime.sbt
 #
 # If you want to add something from your personal setup,
 # add it to your ~/.gitignore instead.
+.idea
```
Comment (fommil) [inline .gitignore]: please follow the instructions.

File: api/src/main/scala/org/ensime/api/config.scala [hunk 9]
```diff
@@ -29,60 +28,33 @@ final case class EnsimeConfig(
...
   def allJars: Set[File] = {
     modules.values.flatMap { m =>
-      m.compileDeps ::: m.testDeps
+      m.libraryJars
     }.toSet
   } ++ javaLibs
 
   val allTargets: Set[File] =
-    subprojects.flatMap(sm => sm.targets ::: sm.testTargets).toSet
+    projects.flatMap(m => m.targets).toSet
```
Comment (fommil) [inline api/src/main/scala/org/ensime/api/config.scala]: you can write `_.targets`

File: api/src/main/scala/org/ensime/api/incoming.scala [hunk 10]
```diff
@@ -53,7 +53,7 @@ final case class TypecheckFileReq(fileInfo: SourceFileInfo) extends RpcAnalyserR
 /**
  * Response with a `VoidResponse`.
  */
-final case class UnloadModuleReq(module: String) extends RpcAnalyserRequest
+final case class UnloadModuleReq(moduleId: EnsimeProjectId) extends RpcAnalyserRequest
```
Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: these changes to the API will show up in the wire protocol. You'll need to add a custom marshaller for `EnsimeProjectId` in jerk(y) and swank(y), e.g. see how we add a custom `DebugThreadIdFormat` (you'll need to add that custom formatter in 3 places, and it wouldn't be a bad idea to add tests)

**Comment by fommil** [comment]
almost there, just a few things. Unfortunately the changes to the wire protocol mean we can't merge just yet.

File: api/src/main/scala/org/ensime/api/incoming.scala (hunk 10 above)
Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: actually, I've changed my mind... let's just change the protocol message. @ktonga isn't this used in vim? This message will change to require a richer message.

In that case, bump the `1.9.1` to `1.9.2` in `outgoing.scala`

Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: or was it @hedefalk who wanted it in atom...

Comment (VC1995) [inline api/src/main/scala/org/ensime/api/incoming.scala]: so should I simply bump the version to 1.9.2 ?

File: .gitignore (hunk 8 above)
Comment (VC1995) [inline .gitignore]: I'm sorry I missed the instructions. I have corrected it.

File: api/src/main/scala/org/ensime/api/incoming.scala (hunk 10 above)
Comment (VC1995) [inline api/src/main/scala/org/ensime/api/incoming.scala]: If we added a custom marshaller we would have the configuration default to `"compile"`.
There could be cases when user might want to typecheck files in the `"test"` configuration as well and might fail later. (say) during resolution related to `ScalaTest` which wasn't among the `libraryJars` in the `"compile"` configuration. Currently since `compileDeps` and `testDeps` are in the same `EnsimeModule` identifiable through `moduleName`it wasn't be a problem. Right?

Comment (ktonga) [inline api/src/main/scala/org/ensime/api/incoming.scala]: @fommil no, we're not using that req in vim ATM

**Comment by fommil** [comment]
cool, could you please just push small commits to a PR with the changes? We can squash when we merge the PR but it makes it easier to see the changes in response to the review.

**Comment by VC1995** [comment]
Yes sure. Sorry about that. I'll take care from now on.

**Comment by fommil** [comment]
first merge, w00t!
